devices: Add fw_cfg device

AlexOrozco1256 opened the pull request:
We are working with the project [Oak](https://github.com/project-oak/oak) team to enable running [Oak Containers](https://github.com/project-oak/oak/tree/main/oak_containers) on cloud hypervisor. These containers are SEV-SNP enlightened and can be used alongside with the rest of the oak suite for attestation and measurement of the cvm. 

Here is the functionality we are adding to CHV.
1. Implmenting the sev-snp APIs previously defined (sev_snp_init, import_isolated_pages, complete_isolated_import). 
2. Expanded the igvm loader to support booting KVM + SEV-SNP enabled VMs. (Note that we can also use the igvm loader to boot VMs with standard cloud images with/without sev-snp enabled as seen here (https://github.com/AlexOrozco1256/chv-demo)
3. Added a fw_cfg device with DMA enabled. Oak's [Stage0](https://github.com/project-oak/oak/tree/main/stage0_bin) firmware uses fw_cfg for loading various parameters into the VM. 

Additional design/implementation details can be found here: https://tinyurl.com/chv-kvm-sev-snp

See: #6653

AlexOrozco1256 commented:
Here is how I tested the fw_cfg device. These are the same steps as the direct kernel boot instructions in the [README](https://github.com/cloud-hypervisor/cloud-hypervisor?tab=readme-ov-file#custom-kernel-and-disk-image) for the chv repo.

The only difference is that we must enable [CONFIG_FW_CFG_SYSFS](https://github.com/cloud-hypervisor/linux/blob/ch-6.12.8/arch/x86/configs/ch_defconfig#L1370) in the cloud hypervisor linux kernel.

Here is how to build and run the vm with fw_cfg device. 
```
git clone --depth 1 https://github.com/cloud-hypervisor/linux.git -b ch-6.12.8 linux-cloud-hypervisor
### ENABLE CONFIG_FW_CFG_SYSFS in arch/x86/configs/ch_defconfig
pushd linux-cloud-hypervisor
make ch_defconfig
KCFLAGS="-Wa,-mx86-used-note=no" make bzImage -j `nproc`
popd

wget https://cloud-images.ubuntu.com/focal/current/oracular-server-cloudimg-amd64.img
qemu-img convert -p -f qcow2 -O raw oracular-server-cloudimg-amd64.img oracular-server-cloudimg-amd64.raw

echo "Hello from fw_cfg!" > /tmp/fw_cfg_test_item

git clone https://github.com/AlexOrozco1256/chv-g.git -b pr-fw_cfg
pushd chv-g
cargo build --features fw_cfg --release
./scripts/create-cloud-init.sh
popd

./chv-g/target/releasecloud-hypervisor \
	--kernel linux-cloud-hypervisor/arch/x86/boot/compressed/vmlinux.bin \
	--disk path=oracular-server-cloudimg-amd64.raw path=/tmp/ubuntu-cloudinit.img\
	--cmdline "console=hvc0 root=/dev/vda1 rw" \
	--cpus boot=4 \
	--memory size=4G \
        --fw_cfg name="opt/org.test/fw_cfg_test_item",file=/tmp/fw_cfg_test_item \
	--net "tap=,mac=,ip=,mask="
```

Now log into the vm with:
User: cloud
Password:cloud123
```
sudo cat /sys/firmware/qemu_fw_cfg/by_name/opt/org.test/fw_cfg_test_item/raw
Hello from fw_cfg!
```

rbradford commented:
@AlexOrozco1256 Please can you rebase to get rid of the merge commit?

AlexOrozco1256 commented:
> @AlexOrozco1256 Please can you rebase to get rid of the merge commit?

Done.

likebreath commented on devices/src/legacy/fw_cfg.rs:
```diff
@@ -42,6 +54,13 @@ struct boot_params {
 // SAFETY: boot_params is only data, reading it from data is a safe initialization.
 unsafe impl ByteValued for boot_params {}
 
+#[cfg(target_arch = "x86_64")]
+const STAGE0_START_ADDRESS: GuestAddress = GuestAddress(0xffe0_0000);
+#[cfg(target_arch = "x86_64")]
+const STAGE0_SIZE: usize = 0x20_0000;
```
Can you please provide some pointers to how these values are defined? Are these specific to Oak firmware? I may miss some background knowledge in general.

likebreath commented on devices/src/legacy/fw_cfg.rs:
```diff
@@ -175,6 +194,63 @@ impl FwCfg {
     }
 
+    pub fn add_e820(&mut self, mem_size: usize) -> Result<()> {
+        #[cfg(target_arch = "x86_64")]
+        let mut mem_regions = vec![
+            (GuestAddress(0), EBDA_START.0 as usize, RegionType::Ram),
+            (
+                MEM_32BIT_DEVICES_START,
+                MEM_32BIT_DEVICES_SIZE as usize,
+                RegionType::Reserved,
+            ),
+            (
+                PCI_MMCONFIG_START,
+                PCI_MMCONFIG_SIZE as usize,
+                RegionType::Reserved,
+            ),
+            (STAGE0_START_ADDRESS, STAGE0_SIZE, RegionType::Reserved),
+        ];
+        #[cfg(target_arch = "aarch64")]
+        let mut mem_regions = arch::aarch64::arch_memory_regions();
+        if mem_size < MEM_32BIT_DEVICES_START.0 as usize {
+            mem_regions.push((
+                HIGH_RAM_START,
+                mem_size - HIGH_RAM_START.0 as usize,
+                RegionType::Ram,
+            ));
+        } else {
+            mem_regions.push((
+                HIGH_RAM_START,
+                MEM_32BIT_RESERVED_START.0 as usize - HIGH_RAM_START.0 as usize,
+                RegionType::Ram,
+            ));
+            mem_regions.push((
+                RAM_64BIT_START,
+                mem_size - (MEM_32BIT_DEVICES_START.0 as usize),
+                RegionType::Ram,
+            ));
+        }
```
We also calculate the e820 maps from couple of other places, say from `configure_pvh` and `configure_32bit_entry` on x86 (not sure what's done on aarch64). I wonder if it would be beneficial to refactor related code so that they can be reused? 

@rbradford You know this area much better than me.

rbradford replied:
Good point but might be tricky as there are lots of very similar but subtly different formats for conveying the same information about e.g. the memory holes.

likebreath reviewed (commented):
Thank you for the good work. Besides comments below, couple other thoughts:

1. Please extend our GH workflows (mostly quality.yaml and build.yaml) to cover the `fw_cfg` feature flag;
2. Please add a documentation to explain what it does and how to use it;

Also, looks like adding `CONFIG_FW_CFG_SYSFS` to the guest kernel config is the only missing bit to add an integration test on `fw_cfg`, as what you described [here](https://github.com/cloud-hypervisor/cloud-hypervisor/pull/7117#issuecomment-2940852628). 

If that's the case, It'd be great to have that integration test added, and I can help with adding the missing kernel config to reference kernel being used by our CI. (This is not required, since we don't require integration test for features behind feature flags).

AlexOrozco1256 commented on vmm/src/vm.rs:
```diff
@@ -716,6 +725,68 @@ impl Vm {
+                    .expect("fw_cfg device must be present")
+                    .lock()
+                    .unwrap()
+                    .add_kernel_data(&kernel_file);
```
I added a FwCfgConfig field to PayloadConfig, we can specify with booleans which of all available parameters we want to pass to the fw_cfg device explicitly.

AlexOrozco1256 commented:
> Thank you for the good work. Besides comments below, couple other thoughts:
> 
> 1. Please extend our GH workflows (mostly quality.yaml and build.yaml) to cover the `fw_cfg` feature flag;
> 2. Please add a documentation to explain what it does and how to use it;
> 
> Also, looks like adding `CONFIG_FW_CFG_SYSFS` to the guest kernel config is the only missing bit to add an integration test on `fw_cfg`, as what you described [here](https://github.com/cloud-hypervisor/cloud-hypervisor/pull/7117#issuecomment-2940852628).
> 
> If that's the case, It'd be great to have that integration test added, and I can help with adding the missing kernel config to reference kernel being used by our CI. (This is not required, since we don't require integration test for features behind feature flags).

I added an integration test that does exactly what my comment [here](https://github.com/cloud-hypervisor/cloud-hypervisor/pull/7117#issuecomment-2940852628) was doing. I use direct kernel boot to boot a vm and verify that the item I added to the fw_cfg device is accessible from within the guest and has the correct contents.

I would just need help from @likebreath to enable the [CONFIG_FW_CFG_SYSFS](https://github.com/cloud-hypervisor/linux/blob/ch-6.12.8/arch/x86/configs/ch_defconfig#L1370) option in the reference kernel.

likebreath commented:
> I would just need help from @likebreath to enable the [CONFIG_FW_CFG_SYSFS](https://github.com/cloud-hypervisor/linux/blob/ch-6.12.8/arch/x86/configs/ch_defconfig#L1370) option in the reference kernel.

Of course. I will try to get to it tomorrow. Thanks for the quick turn-around.

AlexOrozco1256 commented:
> > > I would just need help from @likebreath to enable the [CONFIG_FW_CFG_SYSFS](https://github.com/cloud-hypervisor/linux/blob/ch-6.12.8/arch/x86/configs/ch_defconfig#L1370) option in the reference kernel.
> > 
> > 
> > Of course. I will try to get to it tomorrow. Thanks for the quick turn-around.
> 
> @AlexOrozco1256 Okay. New kernel release was out, plus changes to consume it #7137. I also pushed few commits to kick-off the added fw_cfg tests here. Let's rebase and squash the couple commits into the integration test commit later.
> 
> FWIW, I ran the test locally on a aarch64 machine, and found the test failed. I looked into the guest and found the "/sys/firmware/qemu_fw_cfg/by_name/opt/org.test/fw_cfg_test_item/raw" is missing while "/sys/firmware/qemu_fw_cfg/". I believe the arm64 worker will fail for the same reason. PTAL. Thanks.

I just merged the test commits together and fixed the test_fw_cfg test. 

We have to run cat with sudo for fw_cfg files. I also forgot to add .../raw to the first test. I verified that the test works locally.

likebreath commented:
> I just merged the test commits together 

https://github.com/cloud-hypervisor/cloud-hypervisor/pull/7137 was for kernel version bump which was landed this morning. You should have rebased on main to "adopt" last commit about kernel version bump, rather than quashing it the integration commit. Now I guess you will need to manually resolve some conflicts (which I believe will be not too many).

>and fixed the test_fw_cfg test.
> 
> We have to run cat with sudo for fw_cfg files. I also forgot to add .../raw to the first test. I verified that the test works locally.

Great. I will take another look later this week.

AlexOrozco1256 commented:
> #7137 was for kernel version bump which was landed this morning. You should have rebased on main to "adopt" last commit about kernel version bump, rather than quashing it the integration commit. Now I guess you will need to manually resolve some conflicts (which I believe will be not too many).
> 
Done.
> Great. I will take another look later this week.
Thanks!

AlexOrozco1256 commented:
I fixed the arm test [here](https://github.com/cloud-hypervisor/cloud-hypervisor/compare/935b5bd43a9c483877ee369b71d2c19acf4562cc..14fd76cf34a1b66deb117d5cb4eea1e62bc3d894#diff-7384b95ccd4070a219d1f99e4d163d04522b3a93abb3c303a917f69776525dabR881). I was forgetting to add the fw_cfg device to the device tree which is [required for arm](https://github.com/torvalds/linux/blob/master/drivers/firmware/qemu_fw_cfg.c#L10)

AlexOrozco1256 commented:
Hey Bo, my apologies, I don't have access to an arm server to test this locally. The failing arm test showed "error -16" or device busy which is from the fw_cfg device being presented twice (via acpi and via device tree for arm). 

I fixed this [here](https://github.com/cloud-hypervisor/cloud-hypervisor/compare/c136d4949bbb5227251b179d99c90a4fc7f7c187..3078a66bbcf8de428f5ea599099602f55fedf61f) by target_arch guarding the addition of fw_cfg device to acpi for x86 only. We will present the device via device_tree for arm.

phip1611 commented on vmm/src/vm.rs:
```diff
@@ -335,6 +335,15 @@ pub enum Error {
     #[error("Error locking disk images: Another instance likely holds a lock")]
     LockingError(#[source] DeviceManagerError),
+
+    #[error("Fw Cfg missing kernel file")]
```
nit: `Fw Cfg missing kernel file` -> `Fw Cfg missing kernel`; you also don't have ` file` in the other message

phip1611 commented on vmm/src/vm.rs:
```diff
@@ -335,6 +335,15 @@ pub enum Error {
     LockingError(#[source] DeviceManagerError),
+
+    #[error("Fw Cfg missing kernel file")]
+    FwCfgKernelFile(#[source] io::Error),
```
`FwCfgKernelFile` -> `MissingFwCfgKernelFile`

phip1611 commented on vmm/src/vm.rs:
```diff
@@ -335,6 +335,15 @@ pub enum Error {
+    FwCfgKernelFile(#[source] io::Error),
+
+    #[error("Fw Cfg missing initramfs")]
+    FwCfgInitramfs(#[source] io::Error),
```
`FwCfgKernelFile` -> `MissingFwCfgKernelFile`

phip1611 commented on vmm/src/vm.rs:
```diff
@@ -335,6 +335,15 @@ pub enum Error {
+    FwCfgInitramfs(#[source] io::Error),
+
+    #[error("Fw Cfg missing kernel cmdline")]
+    FwCfgCmdline,
```
`FwCfgCmdline -> `MissingFwCfgCmdline`

phip1611 commented on vmm/src/vm.rs:
```diff
@@ -716,12 +725,72 @@ impl Vm {
-            .map_err(Error::DeviceManager)?;
+        #[cfg(all(feature = "fw_cfg", not(target_arch = "riscv64")))]
+        {
+            device_manager
```
Why not grab the lock of `device_manager` once? I know this pattern is used multiple times in CHV, but I don't think it is optimal

phip1611 commented on devices/src/acpi.rs:
```diff
@@ -3,18 +3,26 @@
 //
 
+#[cfg(all(feature = "fw_cfg", not(target_arch = "riscv64")))]
+use std::mem::{offset_of, size_of};
```
edit, offset_of not, apparently: https://github.com/rust-lang/rust/pull/123168/

phip1611 commented on devices/src/acpi.rs:
```diff
@@ -3,18 +3,26 @@
 use std::time::Instant;
 
+#[cfg(all(feature = "fw_cfg", not(target_arch = "riscv64")))]
+use acpi_tables::rsdp::Rsdp;
```
```rust
#[cfg(all(feature = "fw_cfg", not(target_arch = "riscv64")))]
use {
   acpi_tables::rsdp::Rsdp,
   zerocopy::{FromBytes, Immutable, IntoBytes},
   crate::legacy::fw_cfg::{create_file_name, FwCfgContent, FwCfgItem, FILE_NAME_SIZE},
}
```

phip1611 commented on devices/src/acpi.rs:
```diff
@@ -254,3 +262,204 @@ impl BusDevice for AcpiPmTimerDevice {
+pub const SIGNATURE: [u8; 4] = *b"XSDT";
+pub const COMPILER_ID: [u8; 4] = *b"RVAT";
+
+#[cfg(all(feature = "fw_cfg", not(target_arch = "riscv64")))]
```
there are so many `#[cfg(all(feature = "fw_cfg", not(target_arch = "riscv64")))]`, it might be worth it to have a dedicated module with a single feature gate?

phip1611 commented on vmm/src/acpi.rs:
```diff
@@ -838,6 +842,201 @@ pub fn create_acpi_tables(
+            .contains_key(&(DeviceType::Serial, DeviceType::Serial.to_string()));
+        let serial_device_addr = arch::layout::LEGACY_SERIAL_MAPPED_IO_START.raw_value();
+        let serial_device_irq = if is_serial_on {
+            device_manager
```
same here, we could only grab a single lock of device manager

phip1611 commented on vmm/src/device_manager.rs:
```diff
@@ -1505,6 +1509,31 @@ impl DeviceManager {
         self.bus_devices
             .push(Arc::clone(&fw_cfg) as Arc<dyn BusDeviceSync>);
+
+        // Add a RTC device
```
`Add a RTC device` -> copy & paste?

phip1611 commented on vmm/src/config.rs:
```diff
@@ -163,6 +163,10 @@ pub enum Error {
     #[error("Error parsing --landlock-rules: path/access field missing")]
     ParseLandlockMissingFields,
+    #[cfg(feature = "fw_cfg")]
+    #[error("Error parsing --fw_cfg_items")]
```
Sure it isn't `--fw-cfg-items`? (dashes instead of underscores)

phip1611 commented on vmm/src/config.rs:
```diff
@@ -163,6 +163,10 @@ pub enum Error {
     ParseLandlockMissingFields,
+    #[cfg(feature = "fw_cfg")]
+    #[error("Error parsing --fw_cfg_items")]
+    /// Failed Parsing FwCfgItem config
```
Further, for all the other variants, `#[errors]` follows after the `/// doc comments`: please swap the lines to align style :)

phip1611 commented on vmm/src/config.rs:
```diff
@@ -1603,6 +1613,95 @@ impl FsConfig {
+impl FwCfgConfig {
+    pub const SYNTAX: &'static str = "Boot params to pass to FW CFG device \
+    \"disable<optional; enabled if not set>,e820=on|off,kernel=on|off,cmdline=on|off, \
+    initramfs=on|off,acpi_table=on|off,items=[name0=<backing_file_path>,file0=<file_path>:name1=<backing_file_path>,file1=<file_path>]\"";
```
Are there reasons where 

- one provides `--kernel` and `--cmdline`
- fw_cfg device is used, and
- this information IS NOT passed to fw_cfg?

I'm concerned we might are going fully feature-creep here? If I'm not missing something, I'd like to make this dependend on whether `--kernel` et. al. are avaialble

likebreath replied:
Yes, that's a potential senario as intended. It all depends on what kind of use cases we want to offer to users with `fw_cfg`.

You are right, all booting information needs to be passed to `fw_cfg`, when it is used for booting with oak firmware (Google's main use case for confidential computing) and potentially with native OVMF (we require changes to edk2 to support firmware boot with CH now). 

If you look at how the `fw_cfg` is being used and tested (e.g. the integration test). It is merely for passing a file to the guest, which practically is the only use case we support with this PR.

Until we support and decide to narrow down the `fw_cfg` use case to firmware boot. The current CLI options make more sense to me.

phip1611 commented on vmm/src/vm.rs:
```diff
@@ -728,78 +733,116 @@ impl Vm {
-                .map_err(Error::FwCfgKernelFile)?;
-            if let Some(kernel_file) = kernel {
+        #[cfg(all(feature = "fw_cfg", not(target_arch = "riscv64")))]
+        if fw_cfg_config.is_some() {
```
if let Some(fw_cfg_config) = fw_cfg_config {}

phip1611 reviewed (changes requested):
Thanks for all the work that went into this, good job! I left a few remarks tho.

Further, could you please drop a few words on which boot scenarios will be possible later? I'm referring to #7117

likebreath commented on src/main.rs:
```diff
@@ -269,6 +271,12 @@ fn get_cli_options_sorted(
             .num_args(1..)
             .group("vm-config"),
+        #[cfg(feature = "fw_cfg")]
+        Arg::new("fw_cfg_config")
+            .long("fw_cfg_config")
```
Better rename to fw-cfg-config to be consistent with the rest of CLI.

likebreath commented on vmm/src/vm_config.rs:
```diff
@@ -697,6 +699,72 @@ pub struct PayloadConfig {
+#[cfg(feature = "fw_cfg")]
+#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize)]
+pub struct FwCfgConfig {
+    #[serde(default)]
+    pub enabled: bool,
+    #[serde(default)]
+    pub e820: bool,
+    #[serde(default)]
+    pub kernel: bool,
+    #[serde(default)]
+    pub cmdline: bool,
+    #[serde(default)]
+    pub initramfs: bool,
+    #[serde(default)]
```
I believe `serde` will default to `false`, which is not the behavior with CLI (which default to `true`).

likebreath commented on vmm/src/config.rs:
```diff
@@ -1603,6 +1613,95 @@ impl FsConfig {
+            .0;
+        let items = if parser.is_set("items") {
+            Some(
+                parser
+                    .convert::<FwCfgItemList>("items")
+                    .map_err(Error::ParseFwCfgItem)?
+                    .unwrap(),
```
Can you add a unit test to cover such code path?

likebreath reviewed (commented):
Finally finished my in-depth review of this PR. I believe these should be all my main comments and some questions too. :)

Also, please address the GHA and documentation suggestion that I made from my last [review](https://github.com/cloud-hypervisor/cloud-hypervisor/pull/7117#pullrequestreview-2915181044). 

Thank you again for the good work and persistency.

AlexOrozco1256 commented on devices/src/legacy/fw_cfg.rs:
```diff
@@ -181,6 +206,43 @@ impl FwCfg {
+        #[cfg(target_arch = "aarch64")]
+        let kernel_start = bp.text_offset;
+        #[cfg(target_arch = "x86_64")]
+        let kernel_start = (bp.hdr.setup_sects as usize + 1) * 512;
```
https://docs.kernel.org/arch/x86/boot.html#the-real-mode-kernel-header

setup_sects = 4 is required for backward compat with older linux kernels.
0xff is the wildcard (custom) loader type.

I left comments in the code as well

AlexOrozco1256 commented on devices/src/legacy/mod.rs:
```diff
@@ -22,6 +24,8 @@ mod uart_pl011;
 pub use self::cmos::Cmos;
 #[cfg(target_arch = "x86_64")]
 pub use self::debug_port::DebugPort;
+#[cfg(all(feature = "fw_cfg", not(target_arch = "riscv64")))]
+pub use self::fw_cfg::FwCfg;
```
Will do in a follow-up PR

AlexOrozco1256 commented:
> A separate question: does the fw_cfg device have a feature negotiation mechanism, similar to what virtio uses, to ensure agreement between the guest driver and its implementation from VMM?

[Here](https://www.qemu.org/docs/master/specs/fw_cfg.html#revision-feature-bitmap-key-0x0001-fw-cfg-id) is the feature bitmap defined by the fw_cfg spec 

The only optional feature is the DMA interface which we implemented (here is the [feature bitmap in chv](https://github.com/cloud-hypervisor/cloud-hypervisor/pull/7117/files#diff-f09e9b511d77a1cfc9fdd7eabb8fbe348e724fef49725d622454f543d4c71074R121))

likebreath commented:
> [Here](https://www.qemu.org/docs/master/specs/fw_cfg.html#revision-feature-bitmap-key-0x0001-fw-cfg-id) is the feature bitmap defined by the fw_cfg spec
> 
> The only optional feature is the DMA interface which we implemented (here is the [feature bitmap in chv](https://github.com/cloud-hypervisor/cloud-hypervisor/pull/7117/files#diff-f09e9b511d77a1cfc9fdd7eabb8fbe348e724fef49725d622454f543d4c71074R121))

Right. I see that. May be I miss it, but I don't see how the fw_cfg device negotiates with the guest using such feature bits, say reporting errors when they don't agree. Can you please point me to that?

AlexOrozco1256 commented:
all tests/checks have now been verified on both x86 and aarch64 :) thanks for your patience with all the rebases and thanks a lot @likebreath for helping me facilitate the aarch64 testing

likebreath reviewed (commented):
I am yet to finish the review on the CLI changes, but want to get the comments out early, particularly around ACPI related changes.

Glad to see most of the `fw_cfg` specific code are moved out of the `vmm/acpi.rs` making the changes much less intrusive. My main reservation now is that the`pub fn create_acpi_tables()` is too complicated, since it is generating two set of tables that are very similar with tweaks.

Here is a potential solution: https://github.com/likebreath/cloud-hypervisor/commit/f0836fae4162036fe8758c4d74f4c13b5d21ebcd. Please let me know if it works for `fw_cfg` (particularly if `create_acpi_tables_internal()` can be reused).

AlexOrozco1256 commented:
The last push was a rebase onto main.

I see that other [PRs](https://github.com/cloud-hypervisor/cloud-hypervisor/pull/7208) are also failing the same test, can we try rerunning the CI?

AlexOrozco1256 commented:
https://github.com/cloud-hypervisor/cloud-hypervisor/pull/7117#discussion_r2220329852

I just removed these. These were required to insert the fw_cfg as a device_node!() in device_manager. This macro requires Migrateable which requires the rest.

We added create_fw_cfg_node in fdt.rs which already adds the fw_cfg to the device tree so we don't need this step. I verified that an arm server is still able to find the fw_cfg device without this step by running the integration tests.

likebreath commented on vmm/src/acpi.rs:
```diff
@@ -805,36 +804,108 @@ pub fn create_acpi_tables(
+    for table in &tables {
+        #[cfg(feature = "fw_cfg")]
+        if add_to_fw_cfg {
+            pointers.push(xsdt_offset.0 as usize + xsdt.len());
```
This is quite confusing. Do you intend to do `pointers.push(table)`?

likebreath commented on vmm/src/acpi.rs:
```diff
@@ -626,81 +628,84 @@ fn create_viot_table(iommu_bdf: &PciBdf, devices_bdf: &[PciBdf]) -> Sdt {
+    // To create AcpiTable
+    let mut table_bytes: Vec<u8> = vec![];
+    #[cfg(feature = "fw_cfg")]
+    let mut pointers: Vec<usize> = vec![];
```
IIUC, with `pointers`, what you need is pretty much the offset/address of all tables that XSDT points to, e.g. what `tables` does. Why can't we just reuse `tables`?

I see we need extra address/offset, say `pointer_facp_to_dsdt`, which is fine to be added to `tables` when needed.

likebreath commented on vmm/src/acpi.rs:
```diff
@@ -626,81 +628,84 @@ fn create_viot_table(iommu_bdf: &PciBdf, devices_bdf: &[PciBdf]) -> Sdt {
+    #[cfg(feature = "fw_cfg")]
+    let mut pointers: Vec<usize> = vec![];
+    #[cfg(feature = "fw_cfg")]
+    let mut checksums: Vec<(usize, usize)> = vec![];
```
Maybe the naming is confusing, can you elaborate what exactly that `fw_cfg` needs for creating ACPI tables? I kind of see you need a pair of "offset/address" and "size" for some of the tables.

likebreath reviewed (commented):
Great. The ACPI changes are now much cleaner. I believe we can make `create_acpi_tables_internal()` with less/minimum specifics to `fw_cfg`, while having `create_acpi_tables_for_fw_cfg` handle the specifics. See comments below.

AlexOrozco1256 commented:
As suggested, I modified the code to use the already existing tables to create the other vectors I need (pointer_offsets and checksums) in create_acpi_tables_for_fw_cfg so we don't pass a boolean to create_acpi_tables anymore and there is no additional fw_cfg logic there.

pointer_offsets are the location in the XSDT table where we can find the pointer, so not the actual pointer itself (which is what the pre-existing tables contains)

checksums is a vec of (offset, len) pair used by the firmware when loading the acpi tables to validate each table.

This is essentially an implementation of qemu's bios-linker-loader for acpi tables 
https://github.com/qemu/qemu/blob/master/hw/acpi/bios-linker-loader.c

I left a link and comments where relevant in the code.

likebreath commented on vmm/src/config.rs:
```diff
@@ -1603,6 +1625,105 @@ impl FsConfig {
+#[cfg(feature = "fw_cfg")]
+impl FwCfgConfig {
+    pub const SYNTAX: &'static str = "Boot params to pass to FW CFG device \
+    \"disable<optional; enabled if not set>,e820=on|off,kernel=on|off,cmdline=on|off, \
```
I don't see why we need a `disable` option - The `fw_cfg` device is created when and only when `--fw-cfg-config` option is provided.

likebreath commented on vmm/src/config.rs:
```diff
@@ -2660,6 +2781,23 @@ impl VmConfig {
             }
             disks = Some(disk_config_list);
         }
+        #[cfg(feature = "fw_cfg")]
```
Nit: need a blank line.

likebreath commented on vmm/src/config.rs:
```diff
@@ -2660,6 +2781,23 @@ impl VmConfig {
+            let cmdline = vm_params.cmdline.is_some();
+            let initramfs = vm_params.initramfs.is_some();
+            let fw_cfg_config = FwCfgConfig::parse(vm_params.fw_cfg_config.unwrap())?;
+            if fw_cfg_config.kernel && !kernel {
+                return Err(Error::ParseFwCfgConfigKernel);
+            } else if fw_cfg_config.cmdline && !cmdline {
+                return Err(Error::ParseFwCfgConfigCmdline);
+            } else if fw_cfg_config.initramfs && !initramfs {
+                return Err(Error::ParseFwCfgConfigInitramfs);
+            }
```
Let's separate validate from parse - add a `FWCfgCongfig::validate()` for this and call it from `VmConfig::validate()`.

likebreath commented on vmm/src/vm.rs:
```diff
@@ -733,11 +749,23 @@ impl Vm {
-            .create_fw_cfg_device()
-            .map_err(Error::DeviceManager)?;
+        {
+            let fw_cfg_config = config
+                .lock()
+                .unwrap()
+                .payload
+                .as_ref()
+                .map(|p| p.fw_cfg_config.clone())
+                .unwrap_or_default()
+                .ok_or(Error::VmMissingConfig)?;
```
This is not right - VM will fail to boot without `--fw-cfg-config` option.

likebreath commented on vmm/src/vm_config.rs:
```diff
@@ -699,6 +701,81 @@ pub struct PayloadConfig {
+#[cfg(feature = "fw_cfg")]
+#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize)]
+pub struct FwCfgConfig {
+    pub enabled: bool,
```
Similarly, I don't see the need to have `enabled` filed to `FwCfgConfig`.

likebreath commented on vmm/src/vm.rs:
```diff
@@ -2353,15 +2390,27 @@ impl Vm {
-                &self.numa_nodes,
-                tpm_enabled,
-            );
+            let fw_cfg_config = self
+                .config
+                .lock()
+                .unwrap()
+                .payload
+                .as_ref()
+                .map(|p| p.fw_cfg_config.clone())
+                .unwrap_or_default()
+                .ok_or(Error::VmMissingConfig)?;
```
Ditto - The `--fw-fcg-config` option should not be required to boot a VM.

likebreath commented on vmm/src/vm.rs:
```diff
@@ -798,76 +826,85 @@ impl Vm {
+                .map(|p| p.kernel.as_ref().map(File::open))
+                .unwrap_or_default()
+                .transpose()
+                .map_err(Error::MissingFwCfgKernelFile)?;
```
It should be handled in a central place for the validation of CLI options for the combination of payload. See my comments from #7166. 

I just want to call it out here and let's leave it for future cleanup once #7166 is landed when @phip1611 is back.

likebreath reviewed (commented):
Some comments below around CLI, which should all be relatively easy to resolve.

I believe we are very close to land this large PR, once #7226 is landed (and you can rebase). Thanks again for the persistency.

likebreath commented on vmm/src/acpi.rs:
```diff
@@ -192,6 +192,8 @@ pub fn create_dsdt_table(
     dsdt
 }
 
+pub const FACP_DSDT_OFFSET: usize = 140;
```
Nit: no need to be `pub`.

likebreath commented on vmm/src/acpi.rs:
```diff
@@ -807,6 +809,64 @@ fn create_acpi_tables_internal(
+    ));
+    checksums.push(xsdt_checksum);
+
+    let _ = device_manager
```
Error should be handled properly.

likebreath reviewed (commented):
LGTM. Only two minor comments added. Can you please rebase? I think we are ready to merge.

I was hoping to get the new "linux-loader" release, but that's alright to not include that.

AlexOrozco1256 commented:
> I was hoping to get the new "linux-loader" release, but that's alright to not include that.

I can fix this in a follow-up PR. Thanks for the review! I really appreciate it

AlexOrozco1256 commented:
last push I:
- rebased to get the lychee lint fix https://github.com/cloud-hypervisor/cloud-hypervisor/pull/7247

likebreath reviewed (approved):
Thanks again for the good work and persistency. We are now ready to land it. :)
